### DISCUSSION TITLE
Swap anyhow for thiserror

### DISCUSSION THREAD

**Discussion by wspeirs** [FutureSDR/FutureSDR#23] [Ideas]
Given that FutureSDR is _primarily_ a library, I would recommend switching `anyhow` for `thiserror`. This blog post does a better job at explaining the reasons than I could: https://nick.groenen.me/posts/rust-error-handling/#libraries-versus-applications

The only slightly tricky part would be people creating new blocks that return custom errors for their blocks. This could be handled by simply having the `work` method, for example, return a `WorkError` which simply has an option of `BlockError(Error)` in its enum. The block writer could then create a customer error type, and simply convert to `BlockError` if a problem is encountered.

Thoughts?

**Comment by bastibl** [comment]
Good point. I agree that FutureSDR is mainly a library. And I also have the feeling that it totally overuses `anyhow`. That's probably not only the `work()` function but all over the place. I think it requires a better high-level strategy/concept how to deal with errors. My understanding is that since `work()` is implemented by the user, it would just wrap a `dyn Error` of some kind.

Another option would be that `WorkError` would have types for every in-tree block (like `FftWorkError`, `FirFilterWorkError`, etc). This would then need heavy feature-gating since many blocks are optional. So I don't think that this is nicer. And for out-of-tree blocks it would still require the generic `dyn Error` option.

Also, at the moment, the error doesn't bubble up at the user :-) So it would not show up in applications using FutureSDR, which would be the main point of switching to `thiserror`. So maybe one could keep the generic error for `work()` but have the runtime wrap it in a `WorkError` that annotates more information about the block that caused it and forward it so that it bubbles up in `runtime.run(fg)`.

**Comment by wspeirs** [comment]
It's not too bad, you just have to box up the underlying error. Using `Fft` as an example, create an error type for this block:

```
#[derive(Error, Debug)]
pub enum FftError {
  #[error("Cannot produce FFT")]
  BadFft
}
```

... then you would have an overall `BlockError` type, with an option that takes any error:

```
#[derive(Error, Debug)]
pub enum BlockError {
    #[error("Some kind of internal processing error")]
    InternalError,
    
    #[error("Error happened while doing work: {0}")]
    WorkError(#[from] Box<dyn std::error::Error>)
}
```

... then you'd create and return a `BlockError` as such:

```
return BlockError::WorkError(Box::new(fft_error));
```

Honestly though, I would probably skip all of this, and just make `WorkError` contain a `String`, and any errors from the block can be converted to a `String` and returned. For something like `Block` especially, there _probably_ isn't much the user can do if they cannot compute an `Fft` for example, but it'd be super-helpful for them to know where and why there was a problem.

**Comment by bastibl** [comment]
Yes, what you suggested is what I meant with wrapping a `dyn Error`. Referring to your example, I thought that there may not be any meaningful `InternalError`s, leaving us only with `WorkError`, which just wraps a `dyn Error` and this is already what `anyhow` does (according to my understanding).
But I agree that what you propose is more idiomatic and would be nicer. I'll have to look deeper into Rust error handling.
